pact/exchange.pact

```diff
@@ -354,107 +365,109 @@
-        (numerator (* amountInWithFee reserveOut))
-        (denominator (+ reserveIn amountInWithFee))
+      ( allocs:[object{alloc}]
+        token-out:object{token-path}
+      )
+      (let*
+        ( (head:object{alloc} (at 0 allocs))
+          (p (at 'pair1 token-out))
+          (token-out:object{token-path} token-out)
+          (token-in:object{token-path} (at 'token-out head))
+          (amountIn:decimal (at 'out head))
+          (reserveIn (reserve-for p (at 'token token-in)))
+          (reserveOut (reserve-for p (at 'token token-out)))
+          (amountInWithFee (* (- 1.0 FEE) amountIn))
+          (numerator (* amountInWithFee reserveOut))
+          (denominator (+ reserveIn amountInWithFee))
+        )
+        (+ [ { 'token-out: token-out
+             , 'token-in: token-in
+             , 'in: amountIn
+             , 'out: (truncate (at 'token token-out) (/ numerator denominator))
+             , 'idx: (+ 1 (at 'idx head))
+             , 'pair: p
+             , 'path: (drop 1 (at 'path head))
+             } ]
+           allocs)
       )
-      (+ [ { 'token-out: token-out
-           , 'token-in: token-in
-           , 'in: amountIn
-           , 'out: (truncate token-out (/ numerator denominator))
-           , 'idx: (+ 1 (at 'idx head))
-           , 'pair: p
-           , 'path: (drop 1 (at 'path head))
-           } ]
-         allocs)
     )
-  )
 
 
   (defun swap-exact-out
     ( amountOut:decimal
       amountInMax:decimal
-      path:[module{fungible-v2}]
+      path:[object{token-path}]
       sender:string
       to:string
       to-guard:guard
     )
     (enforce (>= (length path) 2) "swap-exact-out: invalid path")
-    ;; fold over tail of reverse path with dummy first value to compute inputs
-    ;; assembles allocs in forward order
-    (let*
-      ( (rpath (reverse path))
-        (path-len (length path))
-        (pz (get-pair (at 0 rpath) (at 1 rpath)))
-        (e:[module{fungible-v2}] [])
-        (allocs
-          (fold (compute-in)
-            [ { 'token-out: (at 1 rpath)
-              , 'token-in: (at 0 rpath)
-              , 'out: 0.0
-              , 'in: amountOut
-              , 'idx: path-len
-              , 'pair: pz
-              , 'path: e
-              }]
-            (drop 1 rpath)))
-        (allocs1 ;; drop dummy at end, prepend dummy for initial transfer
-          (+ [  { 'token-out: (at 0 path)
-                , 'token-in: (at 1 path)
-                , 'out: (at 'in (at 0 allocs))
-                , 'in: 0.0
-                , 'idx: 0
-                , 'pair: (at 'pair (at 0 allocs))
-                , 'path: path
-             } ]
-             (take (- path-len 1) allocs)))
+      ;; fold over tail of reverse path with dummy first value to compute inputs
+      ;; assembles allocs in forward order
+      (let*
+        ( (rpath (reverse path))
+          (path-len (length path))
+          (pz (at 'pair1 (at 0 rpath)))
+          (e:[object{token-path}] [])
```

Comment:
is the annotation here necessary for runtime type checking or FV? Curious whether this infers/verifies without it.